VDB-1550 Collateral Generator

elizabethengelman opened the pull request.

elizabethengelman committed:
Create a generators directory

elizabethengelman committed:
Add a config file parser for new collateral generator

elizabethengelman committed:
Parse the existing config file

elizabethengelman committed:
Create updated config

- adds flip storage transformerName for new collateral
- adds median storage transformerName for new collateral
- adds flip storage exporter for new collateral
- adds median storage exporter for new collateral
- adds flip, median and osm contracts to event exporters
- adds flip, median and osm contracts

elizabethengelman committed:
Update how config file is parsed

elizabethengelman committed:
Move collateral struct to its own file

elizabethengelman committed:
Rename config generator pkg to config

elizabethengelman committed:
Allow parser and updated to be passed into the generator

- creates IParse and IUpdate interfaces

elizabethengelman committed:
Start to build out test for NewCollateralGenerator

elizabethengelman committed:
Use a different struct for the encoded config

so that we're able to properly annontate the config so that the encoded
toml is formatted correctly

elizabethengelman committed:
Apply go fmt changes

elizabethengelman committed:
Create a new collateral CLI command

elizabethengelman committed:
Handle addContractsToEventExporters error

elizabethengelman committed:
Omit adding new Median and OSM contracts when necessary

This is required for a new collateral like ETH-B where the median and
osm contracts are the same as ETH-A.

elizabethengelman committed:
Update exporter keys created in getTransformerExporters

they should not be prepended with "exporter." because this will be
included automatically when Encoding the config struct, and writing the
toml file.

elizabethengelman committed:
Some small cleanup

elizabethengelman committed:
Update GetUpdatedConfig to return config formatted for toml encoding

elizabethengelman committed:
Ensure that contracts aren't added to event exporters multiple times

elizabethengelman committed:
Update the plugins transformerExporter file

elizabethengelman committed:
Extract out file path helper

elizabethengelman committed:
Restructure collateral generator directory

elizabethengelman committed:
Begin writing flip storage initializers

elizabethengelman committed:
Update toml decoding approach

decode exporters into a map[string]interface to account for dynamic keys
(e.g. exporter.cat_v1_1_0)

elizabethengelman committed:
Create a method to get the updated config formatted for toml encoding

GetUpdatedConfigForToml returns the exporter metadata and transformer
exporters in a single map as as a map[string]interface

elizabethengelman committed:
Use GetUpdatedConfigForToml to write to the config file

elizabethengelman committed:
Write median storage initializer

elizabethengelman committed:
Cleanup & refactoring

elizabethengelman committed:
Switch CLI flags to prompts for collateral generator

elizabethengelman committed:
Ensure that config file keys are written in lowercase

elizabethengelman committed:
Factor out a common createInitializer method

elizabethengelman committed:
Use a different library for cli prompts

paytonrules commented on cmd/newCollateralGenerator.go:
```diff
@@ -0,0 +1,240 @@
+	Use:     "addNewCollateral",
+	Short:   "",
+	Long:    ``,
+	Example: ``,
```
Don't you want an example? I get the lack of a long and short.

elizabethengelman replied:
Nice call - just pushed an update.

elizabethengelman committed:
Add documentation for addNewCollateral cmd

paytonrules commented on cmd/newCollateralGenerator.go:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	configUpdater := config.NewConfigUpdater(collateral, contracts, prompter.MedianRequired, prompter.OsmRequired)
+	configFileName := "mcdTransformers"
```
Is this always mcdTransformers? If so maybe it should be a constant.

rmulhol replied:
agree with the sentiment - would also consider making the config file configurable via CLI flag with this as the default. Trying to think about how we can make it obvious that renaming that file will require a change here 🤔

elizabethengelman replied:
Great point - also could see this being helpful if we ever want to extend this command work for backfill config files.

elizabethengelman committed:
Update go.sum

elizabethengelman committed:
Update initializer generator to use vdb-transformer-utilities

paytonrules commented on generators/new_collateral/config/parser_test.go:
```diff
@@ -0,0 +1,53 @@
+			Expect(parseErr).NotTo(HaveOccurred())
+			Expect(config.Contracts).To(Equal(expectedContracts))
+		})
+	})
```
What about the expected errors?

elizabethengelman replied:
good call! added tests in 938de7a859b2a550b95a0efe656128c314b15f6e

I'd be interested in your thoughts about other ways to test some of the functionality in the parser - the easiest way I could think of was to export some of the previously private methods, but not sure if maybe there's a better way to handle this that I'm not seeing at the moment.

rmulhol commented on cmd/newCollateralGenerator.go:
```diff
@@ -0,0 +1,101 @@
+
+	configUpdater := config.NewConfigUpdater(collateral, contracts, prompter.MedianRequired, prompter.OsmRequired)
+	configFileName := "mcdTransformers"
+	configFilePath := helpers.GetEnvironmentsPath()
```
also wondering if we might want to be able to specify the path, or maybe just loudly warn that this command depends on the config file living in `$GOPATH/src/github.com/makerdao/vdb-mcd-transformers/environments/`

elizabethengelman replied:
makes sense - added in 4ba6331

rmulhol commented on cmd/newCollateralGenerator.go:
```diff
@@ -0,0 +1,101 @@
+		InitializerGenerator: &initializerWriter,
+	}
+
+	fmt.Println(fmt.Sprintf("Adding %s to %s", collateral.Name, helpers.GetFullConfigFilePath(configFileName, configFilePath)))
```
maybe worth using logging facilities here?

elizabethengelman replied:
I had initially used `fmt.Println` for this so that we get some feedback in the command line while running this command, rather than needing to tail the logs to make sure that it's working. But the success/error messages end up being put in the logs, so we need to tail those anyway to see any potential errors. 🤷

elizabethengelman replied:
Actually, just moved some of this code into the generator pkg in 152b8c83f787e7627c451df9695463dee98e8b65, and so as to not print out a lot of output during tests, I've switched to logging vs printing for these.

paytonrules commented on generators/new_collateral/prompts/prompts_wrapper.go:
```diff
@@ -0,0 +1,153 @@
+import (
+	"fmt"
+
+	"github.com/AlecAivazis/survey/v2"
```
Neat!

rmulhol commented on generators/new_collateral/config/parser.go:
```diff
@@ -0,0 +1,85 @@
+type IParse interface {
+	ParseCurrentConfig(configFilePath, configFileName string) (types.TransformersConfig, error)
+}
+type Parser struct{}
```
Really minor, but I wonder if we might want to name the interface `Parser` and then not export the `struct{}` (e.g. make it `parser`) and then use a constructor to return that private implementation

paytonrules replied:
That would be more go-ish wouldn't it?

elizabethengelman replied:
Nice call. 

I was a bit on the fence on wether `parser{}` is really necessary at all since we're not setting any state on it. I decided to put `ParseCurrentConfig` on a parser receiver, so I could create an interface, and use that for testing. But up for other ideas too!

elizabethengelman replied:
addressed in 0018720bd0c7d45029d0e522e0f9e015f202fee8

elizabethengelman replied:
addressed in 0018720bd0c7d45029d0e522e0f9e015f202fee8

rmulhol commented on cmd/newCollateralGenerator.go:
```diff
@@ -0,0 +1,101 @@
+	fmt.Println(fmt.Sprintf("Adding %s to %s", collateral.Name, helpers.GetFullConfigFilePath(configFileName, configFilePath)))
+	addErr := newCollateralGenerator.UpdateConfig()
+	if addErr != nil {
+		return addErr
```
maybe worth decorating the errors with `fmt.Errorf("here's where we failed: %w", err)` throughout to make it clear where we failed if something goes wrong/bubbles up

elizabethengelman replied:
addresses in 663448e6dd0071b7917c35e47b3d5e53cc9071e4

rmulhol commented on generators/new_collateral/initializer/generator.go:
```diff
@@ -0,0 +1,70 @@
+
+var initializerFileName = "initializer.go"
+
+func (g *Generator) GenerateFlipInitializer() error {
```
Do the methods in this file require pointer receivers?

elizabethengelman replied:
good call! updated in ef697ee899f73ccafa9704e28a8b72a1618be8b9

paytonrules commented on generators/new_collateral/test_data/testConfig.toml:
```diff
@@ -0,0 +1,34 @@
+[contract]
+    [contract.MCD_CAT_1_0_0]
+        address  = "0x78f2c2af65126834c51822f56be0d7469d7a523e"
+        abi      = '[{"inputs":[{"internalType":"address","name":"vat_","type":"address"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"bytes32","name":"ilk","type":"bytes32"},{"indexed":true,"internalType":"address","name":"urn","type":"address"},{"indexed":false,"internalType":"uint256","name":"ink","type":"uin ...
```
Do the tests need the entire ABIs?

elizabethengelman replied:
nope, probably not! i can cut them down.  addressed in ef697ee899f73ccafa9704e28a8b72a1618be8b9

rmulhol commented on generators/new_collateral/initializer/generator.go:
```diff
@@ -0,0 +1,70 @@
+	initializer := g.createInitializer(g.Collateral.GetFlipContractName(), "flip")
+	//create the path to the initializer file
+	path := g.Collateral.GetAbsoluteFlipStorageInitializersDirectoryPath()
+	mkDirErr := os.MkdirAll(path, os.ModePerm)
```
could go either way here but I wonder if we might want to consider extracting/wrapping these filesystem IO operations and mocking them in the these tests? Not sure if that gets super complex when calling `Save` on the object returned by `jen.NewFile` 🤔

elizabethengelman replied:
I really like this idea, but may hold off on it for a bit - now that I've added a test for the contents of the initializer, I kind of like being able to see the output of calling `Save` on the `jen.NewFile`. I also think that having a `jen` wrapper may be cool but was struggling to find a good way to do that without needing to mock out each of the render methods, eg `Var()`, `Id()`, `Qual()`, etc.

rmulhol commented on generators/new_collateral/config/parser.go:
```diff
@@ -0,0 +1,85 @@
+	save, saveOk := tomlConfig.Exporter["save"].(bool)
+	schema, schemaOk := tomlConfig.Exporter["schema"].(string)
+	if !homeOk || !nameOk || !saveOk || !schemaOk {
+		return types.ExporterMetaData{}, errors.New("error asserting exporter meta data types")
```
also super minor but I wonder if it'd be helpful to format in the various `ok` values so that we can see what's missing in the error

elizabethengelman replied:
addressed in 0018720

rmulhol commented on generators/new_collateral/config/parser.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	var transformerNames []string
+	decodeErr := mapstructure.Decode(tomlConfig.Exporter["transformerNames"], &transformerNames)
```
do we want to check `tomlConfig.Exporter["transformerNames"]` with a similar `ok` pattern? Not sure if we could run into null pointer errors or silent failures here

elizabethengelman replied:
I just tested it out, and it looks like if `tomlConfig.Exporter["transformerNames"]` were nil, `mapstructure` would just decode it into an empty `[]string`, which I think is ok. Happy to reconsider this, and add a check since we probably don't want an empty slice of transformer names, but OTHO it shouldn't break anything.

elizabethengelman replied:
wrote a test for it in 938de7a859b2a550b95a0efe656128c314b15f6e

rmulhol commented on generators/new_collateral/config/parser.go:
```diff
@@ -0,0 +1,85 @@
+	delete(tomlConfig.Exporter, "name")
+	delete(tomlConfig.Exporter, "save")
+	delete(tomlConfig.Exporter, "schema")
+	delete(tomlConfig.Exporter, "transformerNames")
```
this makes me nervous 😅 

could you help me understand what we're doing here?

paytonrules replied:
Gonna take a whack to make sure I understand - this removes it from an in memory map right?

elizabethengelman replied:
TL;DR: Yep, like Eric said, this piece is removing these keys from an in-memory map. The reason I'm doing that is so it is possible to iterate through the rest of the key-value pairs in this map, which are `transformerExporters`, and then decode those into `TransformerExporter` structs. Since the `transformerExporters` have dynamic keys, it makes it a bit more challenging to auto-decode them as the config file is currently structured.

Longer explanation:
I think that this parser could be cleaned up quite a bit if we change the config file structure - currently we basically have `exporterMetadata` (e.g. "name", "save", "home", etc) and the `transformerExporters` (e.g. exporter.flip_eth_a_v1_1_0, etc) in the same `exporter` map in the config file. But since the `transformerExporter` keys are dynamic, I don't think that we can automatically decode the whole `exporter` map into a struct since the decoder won't know what field to put those `transformerExporters` in. The workaround I'm using in the parser, is to decode the whole `exporter` map into a `map[string]interface{}`, where the values are either the `exporterMetadata` fields, or the `transformerExporters` names.

Then `parseExporterMetaData` and `parseTransformerExporters` are the hacky methods to separate out that generic `exporter` map into two different maps, which can then be automatically decoded into `ExporterMetaData` and `TransformerExporters` structs, which make things ways easier to deal with while we're adding new transformers.

That ended up being longer than I meant it to, and would be happy to chat more about it outside of this PR! 😄

elizabethengelman replied:
Just made a change to just skip the metadata keys, rather than removing them from the map: 110429f714de510edf1148fc4b7d65f7c6772dc0

rmulhol reviewed (commented):
Looking good! Have a few questions, but super excited about where this is headed 🎉 🚀

paytonrules commented on generators/new_collateral/initializer/generator_test.go:
```diff
@@ -0,0 +1,64 @@
+		Expect(removeTestFile).NotTo(HaveOccurred())
+	})
+
+	It("doesn't create a median initializer if it is not configured to do so", func() {
```
Kinda weird there's not a test for the actual contents of the file, but testing strings can suck so 🤷

elizabethengelman replied:
good call - added a test in ce46cfa

paytonrules reviewed (approved):
I have a few questions, but all minor stuff, and this all looks really good.

I'd lean towards merging soon, because even if there are errors we still have to review the code that's generated. A bug isn't likely to cause anything other than inconvenience for us.

elizabethengelman committed:
Address PR feedback

elizabethengelman committed:
Add additional tests for parser

elizabethengelman committed:
Additional PR feedback fixes

elizabethengelman committed:
Extract out a transformer exporter updater

elizabethengelman committed:
Decorate generator errors

elizabethengelman committed:
Refactor ParseTransformerExporters

elizabethengelman committed:
Allow user to set a different directory for config file location

elizabethengelman committed:
Test generated initializer contents

elizabethengelman committed:
Allow for passing in project path to InitializerGenerator

Also removed helper methods that get abosolute initializer paths from
collateral and into IntializerGenerator since that is the only pkg that
uses those methods.

elizabethengelman committed:
Validate deployment block can be converted to an int

elizabethengelman committed:
Apply go fmt changes:

rmulhol commented on generators/new_collateral/test_data/testConfig.toml:
```diff
@@ -0,0 +1,34 @@
+        migrations = "db/migrations"
+        contracts = [
+            "MCD_CAT_1.0.0",
+            "MCD_CAT_1.1.0"
```
I might be missing some nuance but shouldn't the separator be `_` and not `.` here?

elizabethengelman replied:
Yep, this was a copy pasta error I think - it should be `_`. It raises a good point though, as of https://github.com/makerdao/vulcanizedb/pull/147 we now have a way to catch if a `contract.MCD...` entry has contract name formatted incorrectly, but we still don't have a way to catch if we have a list of contracts for a exporter formatted incorrectly.

rmulhol commented on generators/new_collateral/test_data/test_data.go:
```diff
@@ -0,0 +1,357 @@
+		"migrations": CatFileVowExporter.Migrations,
+		"rank":       CatFileVowExporter.Rank,
+		"contracts": []interface{}{
+			"MCD_CAT_1.0.0", "MCD_CAT_1.1.0",
```
Similar question on `_` vs `.` here

rmulhol commented on generators/new_collateral/types/collateral.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (c Collateral) GetOsmContractName() string {
+	// example: MEDIAN_ETH_B
```
:%s/MEDIAN/OSM

rmulhol commented on generators/new_collateral/types/collateral.go:
```diff
@@ -0,0 +1,66 @@
+func (c Collateral) GetFlipContractName() string {
+	// example: MCD_FLIP_ETH_B_1_1_3
+	name := strings.ToUpper(strings.Replace(c.Name, "-", "_", -1))
+	version := fmt.Sprintf("%s", strings.Replace(c.Version, ".", "_", -1))
```
Just making sure I understand - do we need this additional line because these contracts have a version? Will changes be required in this namespace if we add versions to other contracts in the future?

elizabethengelman replied:
Yep, that's right - this was just a way to encapsulate formatting the contract name + it's version. If we decide to add versions to Median and OSM, we'll need to add the collateral version to the `GetOsmContractName` and `GetMedianContractName` methods.

rmulhol commented on generators/new_collateral/initializer/generator.go:
```diff
@@ -0,0 +1,102 @@
+		path := g.GetAbsoluteMedianStorageInitializersDirectoryPath()
+		mkDirErr := os.MkdirAll(path, os.ModePerm)
+		if mkDirErr != nil {
+			return mkDirErr
```
Don't want to go overboard here (don't need to do this as part of this PR) - but I think at least in the medium/long term it'd be cool if wrapped _every_ error in `fmt.Errorf` to clarify exactly where an error bubbled up from if we're logging it and exiting the command

elizabethengelman replied:
👍  planning to do this in a new PR

rmulhol commented on generators/new_collateral/config/updater.go:
```diff
@@ -0,0 +1,278 @@
+	return nil
+}
+
+func (cg *Updater) copyInitialConfig() error {
```
nbd but maybe want to rename the receiver to something other than `cg` now that this is called an `Updater`?

rmulhol commented on generators/new_collateral/config/parser_test.go:
```diff
@@ -0,0 +1,140 @@
+		})
+	})
+
+	Context("ParseExporterMetadata", func() {
```
nbd and probably unnecessary, but is there a reason we can't make these private and test them through `ParseCurrentConfigFile`?

elizabethengelman replied:
👍  planning to take a look at this when working on getting a new PR together

rmulhol reviewed (approved):
LGTM!

Found myself asking questions but I think they're mostly nitpicks and would probably be best addressed in a separate story

elizabethengelman committed:
Address more PR feedback
